tests/test_macro.py

```diff
@@ -13,13 +13,53 @@
 # You should have received a copy of the GNU General Public License
 # along with this program. If not, see <http://www.gnu.org/licenses/>.
 
+import typing
+import pathlib
+
 from unittest.mock import MagicMock, patch
 
 import pytest
 from hamcrest import *
 
 import autokey.service
-import autokey.macro
+from autokey.service import PhraseRunner
+from autokey.configmanager.configmanager import ConfigManager
+import autokey.model
+from autokey.scripting import Engine
+
+from autokey.macro import *
+
+
+def create_engine() -> typing.Tuple[Engine, autokey.model.Folder]:
+    # Make sure to not write to the hard disk
+    test_folder = autokey.model.Folder("Test folder")
+    test_folder.persist = MagicMock()
+
+    # Mock load_global_config to add the test folder to the known folders. This causes the ConfigManager to skip it’s
+    # first-run logic.
+    with patch("autokey.model.Phrase.persist"), patch("autokey.model.Folder.persist"),\
+         patch("autokey.configmanager.configmanager.ConfigManager.load_global_config",
+               new=(lambda self: self.folders.append(test_folder))):
+        engine = Engine(ConfigManager(MagicMock()), MagicMock(spec=PhraseRunner))
+        engine.configManager.config_altered(False)
+
+    return engine, test_folder
+
+def test_arg_parse():
```

Comment:
With this kind of structural regularity in the test inputs, you can parametrize the test and save the typing/copy&pasting the same assert statement over and over again.
just add 3 arguments to the test case function like `test_input, expected, error_msg` and put
the `@pytest.mark.parametrize` decorator in the front of the definition.
Like
```python
@pytest.mark.parametrize("test_input, expected, error_msg", [
    ("<script name='test name' args='long arg with spaces and ='>", {"name": "test name", "args": "long arg with spaces and ="}, "Macro arg can't contain equals"),
…
])
def test_arg_parse(test_input, expected, error_msg):
    engine, folder = create_engine()
    macro = ScriptMacro(engine)
    assert_that(macro._get_args(test_input), is_(equal_to(expected)),
                error_msg)
```
Then adding a new case is just adding a new tuple with test data.

Reply:
Thank you! I saw that in one of the other test files, but didn't understand how it was used.

Reply:
In those, I went a step further. You can pass any iterable, like a Generator function instead of a list into `@pytest.mark.parametrize()`.

So I wrote functions named like `generate_test_cases_for_<insert_test_name_here>()` and placed them in front of the test function. Then I used some nested helper functions inside the generator functions to cut down typing redundant bits and used those to `yield` one test case per line.

This is quite advanced but makes it super easy to add additional test cases to existing tests.

Reply:
I'll push this change to `macro` branch, pull #340

Reply:
Any suggestions for how to parametrize `test_engine.py`'s `create_phrase`? Since it needs a `Folder` object created, but doesn't always use it?

Reply:
Partition the test cases by the used parameters and the expected result (positive/success, negative/raisedException). Each partition of the tests gets it’s own function. Then parameters can be used to raise the number of tests by testing multiple valid/invalid inputs.

Reply:
But how do you pass an object arg instead of a string without making it a global arg?

Reply:
But how do you pass an object arg instead of a string without making it a global arg?

Reply:
> But how do you pass an object arg instead of a string without making it a global arg?

Sorry, it seems that I missed your last question.

Normally, you can pass anything as an argument to the test function.
But If you create class objects of any AutoKey class inside the test parameter list, any Exceptions caused during the object creation won’t be associated to the failing test case. 

Instead, those Exceptions will be risen during the test collection phase (because the test case generators are run beforehand). And Exceptions during the collection phase cause more things to fail.

I’m not quite sure, but it may even kill the whole test suite with an "Errors occurred during the collection phase." error. At least, it fails the whole test.

If we start to use `pyfakefs` to test file system interactions, the `Path` instances must be created inside the `test_function()`, otherwise it will point to the outside world. There may be similar issues with other packages.

That’s why I’m restricting myself to use built-in types or simple structs of built-in types only.